casper/src/main/scala/coop/rchain/casper/util/rholang/costacc/PreChargeDeploy.scala

```diff
@@ -31,14 +29,12 @@ final class PreChargeDeploy(chargeAmount: Long, pk: PublicKey, rand: Blake2b512R
       (`sys:casper:return` ->> GUnforgeable) :: HNil
 
   import toPar._
-  protected override val toEnvMap = ToEnvMap[Env]
+  protected override val envsReturnChannel = Contains[Env, `sys:casper:return`]
```

Comment:
Can we de-duplicate `envsReturnChannel` from `PreChargeDeploy` and `RefundDeploy` by defining it in `SystemDeploy` using the `Env` existential?

Reply:
Unfortunately, I don't think it would work. In `SystemDeploy` `Env` is, as you noted, existential which means that it's structure is not known. When you take a look at the implicit expansion at, say, `PreChargeDeploy` you will notice that the expression `Contains[Env, returnChannel]` expands to `NormalizerEnv.Contains.Aux[(sys:casper:deployerId ->> GDeployerId) :: (sys:casper:invalidBlockHash ->> Expr.ExprInstance.GString) :: (sys:casper:return ->> GUnforgeable) :: HNil,sys:casper:return,ev.KeyType]` - this is precisely the info we're missing in `SystemDeploy` . I don't see how it would work :disappointed:

Reply:
PS. I guess my point is that even though the expression looks the same, it actually expands to different things (using knowledge present at the point of expansion).